### ISSUE TITLE
why there is no need to surround the normal samples with bos and eos or eot like the fim samples when finetuning?

### ISSUE THREAD

**Issue by allenliu88** [BohdanPetryshyn/code-llama-fim-fine-tuning#7]
why there is no need to surround the normal samples with bos and eos or eot like the fim samples?

## munully set the add_special_tokens=False, will not surround the bos and eos automatically
            tokenized_inputs = self.tokenizer(
                buffer, truncation=False, add_special_tokens=False
            )["input_ids"]
            tokenized_inputs = functools.reduce(
                lambda x, y: np.concatenate([x, [self.concat_token_id], y]),
                tokenized_inputs,
            )

by the way, when we use the codellama for text completion, the official inference scripts will be add the bos token before the input text, please ref. https://github.com/meta-llama/codellama/blob/main/llama/generation.py#L220

## the bos = True
prompt_tokens = [self.tokenizer.encode(x, bos=True, eos=False) for x in prompts]

In my opinion, it should be the same processing as finetuning phase and inferencing phase? but i'an not sure!

**Comment by BohdanPetryshyn** [comment]
Hello @allenliu88, thank you for your question! During the implementation, I aimed to closely follow the techniques described in the “Efficient Training of Language Models to Fill in the Middle” and “Code Llama: Open Foundation Models for Code” papers. However, I recall that this specific implementation detail was somewhat unclear. I experimented with various approaches and found that this one yielded the best results, so I assumed it was correct.

I encourage you to modify the implementation by adding the BOS token and share your findings here. You might achieve better results and enhance the implementation which would be a great contribution 🙏

**Comment by allenliu88** [comment]
thank you for your reply @BohdanPetryshyn .

i'm pleasant to try it again! Previously, i also use the same way as you to finetune the deepseek-coder-6.7b-base, which is the same architecture as the code llama, without the bos token ahead of the normal samples, the performance is ok actually. and i will spend some time to revalidate the bos format and try it again in normal samples.

---

i have another question need your help!

Now, after the above finetuning(or sometimes call continuing training) phase, i try to use SFTTrainer to finetune the base model on high quality private FIM datasets, then comparing the differences of this two performance. after the huggingface trl SFTTrainer.__init__ processing, i'm sure that the final processed samples as follow(not packing, no padding) are the same as the previous fim training samples(the difference is no packing, no padding). **my problem encountered is: after the SFT training, even though the finetuned model has learned som new knowledge, but the model cannot predict the <｜end▁of▁sentence｜> stop tokens properly when fim code infilling, it firstly output the correct code result and at the end it just repeat the final output infinitely, at last to exceed the maximum new tokens limit, do you have any ideas about this phenomenon**?

Row 0: raw text
Text: 
```
<｜fim▁begin｜>@Override
    public void visit(Select select) {
        <｜fim▁hole｜>
            }
        }
        select.getSelectBody().accept(this);

    }<｜fim▁end｜>if (select.getWithItemsList() != null) {
            for (WithItem withItem : select.getWithItemsList()) {
                withItem.accept(this);<｜end▁of▁sentence｜>
```

First row: after the SFTTrainer.__init__ processing:
{'input_ids': [32013, 32016, 3566, 7050, 4926, 841, 7459, 3790, 7, 2005, 3346, 3790, 8, 507, 185, 436, 437, 13, 4878, 3790, 405, 3346, 3790, 26, 185, 294, 32015, 185, 315, 611, 32017, 2125, 437, 26, 32014], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
['<｜begin▁of▁sentence｜>', '<｜fim▁begin｜>', 'public', 'ĠFile', 'Config', 'Ġset', 'Default', 'Key', '(', 'String', 'Ġdefault', 'Key', ')', 'Ġ{', 'Ċ', 'ĠĠĠĠĠĠĠ', 'Ġthis', '.', 'default', 'Key', 'Ġ=', 'Ġdefault', 'Key', ';', 'Ċ', 'ĠĠĠĠĠĠĠĠ', '<｜fim▁hole｜>', 'Ċ', 'ĠĠĠ', 'Ġ}', '<｜fim▁end｜>', 'return', 'Ġthis', ';', '<｜end▁of▁sentence｜>']

**Comment by BohdanPetryshyn** [comment]
I haven't faced this issue myself, so I'm not entirely sure what's causing it. A couple of things that come to mind:

There might be a discrepancy in how special tokens are used during the original training, fine-tuning, and inference stages.
It's possible that the fine-tuning process introduced a bias towards generating very long infills.

**Comment by allenliu88** [comment]
got it, thank you very much!
